### PR TITLE
SALTO-4764 - Zendesk Adapter- Prevent removal of dynamic content that is being used

### PR DESCRIPTION
**Pull request by idoser96**
prevent removal of dynamic content that is being used by triggers, macros and automations

---

None

---
_Release Notes_: 
Zendesk Adapter:
* Prevent removal of dynamic_content that is being used by types that blocks its deletion

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * Prevent deletion of dynamic_content if it is used in trigger, macro or automation
+ */
+export const dynamicContentDeletionValidator: ChangeValidator = async (changes, elementSource) => {
+  if (elementSource === undefined) {
+    log.error('Failed to run dynamicContentDeletionValidator because element source is undefined')
+    return []
+  }
+
+  const dynamicContentRemovals = changes.filter(isInstanceChange).filter(isRemovalChange).map(getChangeData)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts:49]: you didn't filter by dynamic content

File: packages/zendesk-adapter/test/change_validators/dynamic_content_deletion.test.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        detailedMessage: 'This dynamic content is being used and cannot be deleted, used by: zendesk.automation.instance.automation, zendesk.macro.instance.macro, zendesk.trigger.instance.trigger',
+      },
+    ])
+  })
+  it('should not error if the user of the dynamic content is removed or modified without the dynamic content', async () => {
+    const changes = [toChange({ before: dynamicContent })]
+    trigger.value = {}
+    const elementsSource = elementSource.createInMemoryElementSource([dynamicContent, trigger, macro, automation])
+    const errors = await dynamicContentDeletionValidator(changes, elementsSource)
+    expect(errors).toMatchObject([
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/change_validators/dynamic_content_deletion.test.ts:86]: the test checks for an error when the description says there shouldn't be an error

**Review by edenhassid — COMMENTED** [review]
looks good! 🍻  added some comments

File: packages/zendesk-adapter/test/change_validators/dynamic_content_deletion.test.ts (hunk 2 above)
Comment (idoser96) [reply packages/zendesk-adapter/test/change_validators/dynamic_content_deletion.test.ts:88]: Added an explaination

File: packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    log.error('Failed to run dynamicContentDeletionValidator because element source is undefined')
+    return []
+  }
+
+  const dynamicContentRemovals = changes
+    .filter(isInstanceChange)
+    .filter(isRemovalChange)
+    .map(getChangeData)
+    .filter(instance => instance.elemID.typeName === DYNAMIC_CONTENT_ITEM_TYPE_NAME)
+  const relevantTypesInstances = await getInstancesFromElementSource(elementSource, TYPES_WITH_DC)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts:54]: can you check if `dynamicContentRemovals === undefined` and return?

**Review by edenhassid — APPROVED** [review]
looks good! 🌴 make sure you address the comment before you merge 😸

File: packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts (hunk 3 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts:54]: i guess you means dynamicContentRemovals.length === 0

Comment (edenhassid) [reply packages/zendesk-adapter/src/change_validators/dynamic_content_deletion.ts:59]: yes

**Pull request merged**